src/tally.cpp

```diff
@@ -420,34 +448,103 @@ std::string pyne::Tally::mcnp(int tally_index, std::string mcnp_version) {
-                                entity_id, entity_size, normalization);
+      tally_id = 2;
     }
+    output << form_mcnp_tally(tally_index, tally_id, particle_token, entity_id,
+                              entity_size, normalization);
 
   } else if (entity_type.find("Volume") != std::string::npos) {
     if (tally_type.find("Flux") != std::string::npos) {
-      output << form_mcnp_tally(tally_index, 4, particle_token, 
-                                entity_id, entity_size, normalization);
+      tally_id = 4;
     } else if (tally_type.find("Current") != std::string::npos) {
       // makes no sense in mcnp
+      return "";
+    }
+    output << form_mcnp_tally(tally_index, tally_id, particle_token, entity_id,
+                              entity_size, normalization);
+  } else if (entity_type.find("Mesh") != std::string::npos) {
+    output << "FMESH" << tally_index << "4:" << particle_token << "  ";
+    output << "GEOM=";
+    std::string indent_block = "           ";
+    std::stringstream sup_var;
+
+    if (entity_geometry.find("Cartesian") != std::string::npos) {
+      output << "XYZ ";
+    } else if (entity_geometry.find("Cylinder") != std::string::npos) {
+      output << "CYL ";
+      if (!is_zero(axl)) {
+        sup_var << indent_block << "AXL=" << to_string(axl);
+      }
+      if (!is_zero(vec)) {
+        sup_var << "\n" << indent_block;
+        sup_var << "VEC=" << to_string(vec);
+      }
+    }
+
+    output << " ORIGIN=" << to_string(origin) << "\n";
+    std::string dir_name[3] = {"I", "J", "K"};
+    std::vector<double> meshes[3] = {i_meshs, j_meshs, k_meshs};
+    std::vector<int> bins[3] = {i_bins, j_bins, k_bins};
+
+    for (int j = 0; j < 3; j++) {
+      output << indent_block;
+      output << dir_name[j] << "MESH=" << to_string(meshes[j]);
+      if (bins[j].size() > 0) {
+        output << " " << dir_name[j] << "INTS=" << to_string(bins[j]);
+      }
+      output << "\n";
+    }
+    if (sup_var.str().size() > 0) {
+      output << sup_var.str() << "\n";
+    }
+    if (e_bounds.size() > 0) {
+      output << indent_block << "EMESH=";
+      output << to_string(e_bounds);
+    }
+    output << "\n";
+    if (e_bins.size() > 0) {
+      output << indent_block << "EINTS=";
+      output << to_string(e_bins);
+    }
+    if (out.size() > 0) {
+      output << "\n" << indent_block << "OUT=" << out;
     }
   } else {
-    std::cout << "tally/entity combination makes no sense for MCNP" << std::endl;
+    std::cout << "tally/entity combination makes no sense for MCNP"
+              << std::endl;
   }
-  
-
   // print sd card if area/volume specified
   return output.str();
-} 
+}
+template <typename T>
+bool pyne::Tally::is_zero(T vect) {
+  int size = sizeof(vect) / sizeof(vect[0]);
+  bool result = true;
+  for (int i = 0; i < size; i++) result &= (vect[i] == 0);
+  return result;
+}
+
+template<typename T>
+std::string pyne::Tally::mcnp_list_to_string(std::vector<T> vect){
```

Comment:
I am wondering is this specific to mcnp ? or could it is used in other cases ?

Shall it go to utils ?

Reply:
cc @gonuke

Reply:
It could be more general, but the key is that it is space delimited.  If you added an argument that was a string to be the delimiter, then it is likely more general.  You might name it `join_to_string` or something

Reply:
I try to move this in `utils.cpp/utils.h`... but it won't compile (on macOS) if I keep the lib as shared...
@makeclean I remember you help me fix a similar issue on cyclus, to you have any idea how to fix it ?